README.md
Page for Custom Post Type
=========================

Specify a page to use for the base URL of a custom post type via your WordPress Reading Settings admin page.
Works in a similar way to how you can set a page for posts in the WordPress reading settings.

On a post type archive page, if your theme's archive template uses the following recommended teplate tags then your page content will be output in place:

 - `the_archive_title()` : The page title
 - `the_archive_description()` : The page content

This plugin may not work with post types that specify their own custom permalink structure.

Installation
------------

1. Upload the `page-for-cpt` folder to the `/wp-content/plugins/` directory.
2. Activate the plugin through the 'Plugins' menu in WordPress.
3. Visit `Admin > Settings > Reading` in the WordPress admin to set which page to use for each post type.


Changelog
---------

View a list of all plugin changes in [CHANGELOG.md].

[CHANGELOG.md]: https://github.com/benhuson/page-for-cpt/blob/master/CHANGELOG.md


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [0.6] - 2018-04-25

### Added
- Add "Edit Page" toolbar item when viewing a post type archive with associated page.
- Add "{Post Type} Page" label to archive pages in admin pages list view.

### Fixed
- Get the correct post type on archive pages, even if there are no posts.
- Automatically flush rewrite rules after settings are changed.

## [0.5] - 2017-04-27

### Fixed
- Add the wp-editor back into the page for posts after it was removed in WordPress 4.2.2.

## [0.4] - 2016-09-15

### Added
- Add and use `get_public_post_types()` method.
- Add and use `get_page_for_cpt_option()` method.
- Add support for `Page_For_CPT::get_the_archive_title()` and `Page_For_CPT::get_the_archive_description()` on post archives.

### Changed
- Init all hooks after plugins are loaded.
- Remove `TEXTDOMAIN` as language domains must be a string not a variable.
- Deprecate and remove plugin directory and url path constants and remove with class methods.
- Use `register_post_type_args` hook if WordPress 4.4+
- Don't filter built-in post type args.
- Only load admin functionality in admin.

## [0.3] - 2015-02-26

### Added
- Filter archive title and description to show page title and content.
- When settings are changed, flush rewrite rules.
- Style admin fields in a table.
- Use PAGE_FOR_CPT_TEXTDOMAIN constant.

### Changed
- Do not require the `page_for_cpt_slug` filter. Automaticatically apply to registered post types.
- Change default admin field option to "Not Set".

## [0.2] - 2015-01-16

### Added
- Added `Page_For_CPT::get_post_type_for_page()` method.
- Added `Page_For_CPT::get_page_for_post_type()` method.

## [0.1] - 2014-09-30

### Added
- First public release.
- Add `page_for_cpt_use_page_body_classes` filter to use page body classes instead of archive body classes.

[Unreleased]: https://github.com/benhuson/page-for-cpt/compare/0.6...HEAD
[0.6]: https://github.com/benhuson/page-for-cpt/compare/0.5...0.6
[0.5]: https://github.com/benhuson/page-for-cpt/compare/0.4...0.5
[0.4]: https://github.com/benhuson/page-for-cpt/compare/0.3...0.4
[0.3]: https://github.com/benhuson/page-for-cpt/compare/0.2...0.3
[0.2]: https://github.com/benhuson/page-for-cpt/compare/0.1...0.2
[0.1]: https://github.com/benhuson/page-for-cpt/releases/tag/0.1


admin/admin.php
<?php

/**
 * @package     Page for CPT
 * @subpackage  Admin
 */

if ( ! defined( 'ABSPATH' ) ) exit;  // Exit if accessed directly.

add_action( 'admin_init', array( 'Page_For_CPT_Admin', 'init_settings_field' ), 500 );
add_action( 'edit_form_after_title', array( 'Page_For_CPT_Admin', 'fix_no_editor_on_posts_page' ), 0 );
add_filter( 'display_post_states', array( 'Page_For_CPT_Admin', 'display_post_states' ), 8, 2 );
add_action( 'update_option_page_for_cpt', array( 'Page_For_CPT_Admin', 'option_updated' ), 10, 3 );
add_action( 'admin_init', array( 'Page_For_CPT_Admin', 'flush_permalinks' ) );

class Page_For_CPT_Admin {

	/**
	 * Init Settings Field
	 *
	 * @since     0.4
	 * @internal  Called by the `admin_init` hook.
	 */
	public static function init_settings_field() {

		add_settings_field(
			'page_for_cpt',
			__( 'Pages for Post Types', 'page-for-cpt' ),
			array( get_class(), 'setting_field' ),
			'reading',
			'default'
		);

		register_setting( 'reading', 'page_for_cpt', array( get_class(), 'validate_setting_field' ) );

	}

	/**
	 * Setting Field
	 *
	 * @since     0.4
	 * @internal  Called by the `init_settings_field()`.
	 */
	public static function setting_field() {

		$page_for_cpt = Page_For_CPT::get_page_for_cpt_option();

		$post_types = Page_For_CPT::get_public_post_types();

		if ( count( $post_types ) > 0 ) {

			echo '<table class="page-for-cpt">';

			foreach ( $post_types as $post_type ) {

				// Skip invalid post types
				if ( ! post_type_exists( $post_type ) ) {
					continue;
				}

				$post_type = get_post_type_object( $post_type );
				$selected = Page_For_CPT::get_page_for_post_type( $post_type->name );

				$id = 'page_for_cpt_' . sanitize_key( $post_type->name );

				?>

				<tr>
					<td>
						<label for="<?php echo $id; ?>"><?php echo $post_type->label; ?>:</label>
					</td>
					<td>
						<?php

						wp_dropdown_pages( array(
							'id'                => $id,
							'name'              => 'page_for_cpt[' . $post_type->name . ']',
							'show_option_none'  => sprintf( '&mdash; %s &mdash;', _x( 'Select', 'menu option', 'page-for-cpt' ) ),
							'option_none_value' => '0',
							'selected'          => $selected
						) );

						?>
					</td>
				</tr>

				<?php

			}

			echo '</table>';
			echo '<style>table.page-for-cpt td { padding: 0 20px 0.5em 0; }</style>';

		} else {

			?>

			<span class="description"><?php _e( '(No custom post types available)', 'page-for-cpt' ); ?></span>

			<?php

		}

	}

	/**
	 * Validate Setting Field
	 *
	 * When settings are changed, flush rewrite rules.
	 *
	 * @since     0.4
	 * @internal  Called by the `init_settings_field()`.
	 *
	 * @param   string  $value  Value.
	 * @return  string          Validated value.
	 */
	public static function validate_setting_field( $value ) {

		$page_for_cpt = Page_For_CPT::get_page_for_cpt_option();

		if ( $value !== $page_for_cpt ) {
			flush_rewrite_rules( true );
		}

		return $value;

	}

	/**
	 * Display Post States
	 *
	 * @since     0.6
	 * @internal  Called by the `display_post_states` filter.
	 *
	 * @param   array    $post_states  Post states.
	 * @param   WP_Post  $post         Post object.
	 * @return  array                  States.
	 */
	public static function display_post_states( $post_states, $post ) {

		$page_for_cpt = Page_For_CPT::get_page_for_cpt_option();
		$post_type = array_search( $post->ID, $page_for_cpt );

		// Is post type page
		if ( ! empty( $post_type ) ) {
			$post_type_object = get_post_type_object( $post_type );
			$post_states[ 'page_for_cpt' ] = sprintf( esc_html__( '%s Page', 'page-for-cpt' ), esc_html( $post_type_object->labels->name ) );
		}

		return $post_states;

	}

	/**
	 * Fix No Editor On Posts Page
	 * 
	 * Add the wp-editor back into WordPress after it was removed in 4.2.2.
	 *
	 * @since     0.5
	 * @internal  Private. Called via the `edit_form_after_title` action.
	 *
	 * @see  /wp-admin/edit-form-advanced.php
	 *
	 * @param  WP_Post  $post  Post object.
	 */
	public static function fix_no_editor_on_posts_page( $post ) {

		if ( $post->ID != get_option( 'page_for_posts' ) ) {
			return;
		}

		remove_action( 'edit_form_after_title', '_wp_posts_page_notice' );
		add_post_type_support( 'page', 'editor' );

	}

	/**
	 * Option Updated
	 *
	 * Set flag the permalink rewrite rules need flushing.
	 *
	 * @since     0.6
	 * @internal  Private. Called via the `update_option_page_for_cpt` action.
	 *
	 * @param  string  $old_value  Old value.
	 * @param  string  $value      New value.
	 * @param  string  $option     Option.
	 */
	public static function option_updated( $old_value, $value, $option ) {

		if ( $old_value !== $value ) {
			update_option( 'page_for_cpt_requires_flush', 1 );
		}

	}

	/**
	 * Flush Permalinks
	 *
	 * Flush rewrite rules if flag set.
	 *
	 * @since     0.6
	 * @internal  Private. Called via the `admin_init` action.
	 */
	public static function flush_permalinks() {

		if ( 1 == get_option( 'page_for_cpt_requires_flush', 1 ) ) {
			flush_rewrite_rules();
			update_option( 'page_for_cpt_requires_flush', 0 );
		}

	}

}


includes/menu.php
<?php

/**
 * @package     Page for CPT
 * @subpackage  Menu
 */

class Page_For_CPT_Menu {

	/**
	 * Set up menu filters.
	 */
	public function __construct() {

		add_filter( 'nav_menu_css_class', array( $this, 'nav_menu_css_class' ), 10, 4 );

	}

	/**
	 * Filters the CSS class(es) applied to a menu item's list item element.
	 *
	 * This filter adds `parent` and `ancestor` classes to any page menu item
	 * that acts as an archive for the current single post type page.
	 *
	 * We treat the archive page as a `parent` if the current post page does
	 * not have a parent post, and an as `ancestor` if the post is the child
	 * of another post (hierarchical).
	 *
	 * @since     WP 3.0.0
	 * @since     WP 4.1.0  The `$depth` parameter was added.
	 * @internal  Called via the `nav_menu_css_class` filter.
	 *
	 * @param   array   $classes  The CSS classes that are applied to the menu item's `<li>` element.
	 * @param   object  $item     The current menu item.
	 * @param   array   $args     An array of wp_nav_menu() arguments.
	 * @param   int     $depth    Depth of menu item. Used for padding.
	 * @return  array             CSS classes.
	 */
	public function nav_menu_css_class( $classes, $item, $args, $depth ) {

		// Only check page menu items
		if ( is_singular() && $this->is_page_menu_item( $item ) ) {

			$post_obj = get_queried_object();

			$post_type = $this->get_current_post_type();
			$page_id = Page_For_CPT::get_page_for_post_type( $post_type );
			$page_ancestor_ids = get_ancestors( $page_id, 'page' );

			// If the page menu item is an archive page, add classes
			if ( $page_id > 0 && $item->object_id == $page_id ) {

				// Parent or ancestor?
				$relation = $post_obj->post_parent > 0 ? 'ancestor' : 'parent';

				$classes = $this->add_archive_page_menu_item_classes( $classes, $post_type, $relation );

			} elseif ( ! empty( $page_ancestor_ids ) && in_array( $item->object_id, $page_ancestor_ids ) ) {

				$classes = $this->add_archive_ancestor_page_menu_item_classes( $classes );

			}

		}

		if ( is_post_type_archive() && $this->is_page_menu_item( $item ) ) {

			$post_type = $this->get_current_post_type();
			$page_id = Page_For_CPT::get_page_for_post_type( $post_type );
			$page_ancestor_ids = get_ancestors( $page_id, 'page' );

			// If the page menu item is an archive page, add classes
			if ( $page_id > 0 && $item->object_id == $page_id ) {

				$classes = $this->add_page_menu_item_classes( $classes );

			} elseif ( ! empty( $page_ancestor_ids ) && in_array( $item->object_id, $page_ancestor_ids ) ) {

				$classes = $this->add_archive_ancestor_page_menu_item_classes( $classes );

			}

		}

		return $classes;

	}

	/**
	 * Menu Item Is Page
	 *
	 * @param   WP_Post  $item  Menu item object.
	 * @return  boolean
	 */
	private function is_page_menu_item( $item ) {

		return 'post_type' == $item->type && 'page' == $item->object;

	}

	/**
	 * Get Current Post Type
	 *
	 * @return  string  Post type.
	 */
	private function get_current_post_type() {

		$qo = get_queried_object();

		if ( is_post_type_archive() ) {

			return $qo->name;

		} elseif ( is_singular() ) {

			return get_post_type( $qo );

		}

		return '';

	}

	/**
	 * Add Page Menu Item Classes
	 *
	 * @param   array    $classes  Menu item classes.
	 * @param   WP_Post  $post     Post object.
	 * @return  array              Classes.
	 */
	private function add_page_menu_item_classes( $classes ) {

		$classes[] = 'current-menu-item';

		return array_unique( $classes );

	}

	/**
	 * Add Archive Page Menu Item Classes
	 *
	 * @param   array           $classes  Menu item classes.
	 * @param   WP_Post|string  $post     Post object.
	 * @param   integer         $parent   Post parent.
	 * @return  array                     Classes.
	 */
	private function add_archive_page_menu_item_classes( $classes, $post_type = '', $relation = '' ) {

		// Sanitize class vars
		$post_type = sanitize_html_class( $post_type );
		$relation = sanitize_html_class( $relation );

		// Add `parent` or `ancestor` menu classes
		$classes[] = 'current-menu-' . $relation;
		$classes[] = 'current-post_type-' . $relation;
		$classes[] = 'current-' . $post_type . '-' . $relation;

		return array_unique( $classes );

	}

	/**
	 * Add Archive Ancestor Page Menu Item Classes
	 *
	 * @param   array  $classes  Menu item classes.
	 * @return  array            Classes.
	 */
	private function add_archive_ancestor_page_menu_item_classes( $classes ) {

		// Add page `ancestor` menu classes
		$classes[] = 'current-menu-ancestor';
		$classes[] = 'current-page-ancestor';

		return array_unique( $classes );

	}

}


page-for-cpt.php
<?php

/*
Plugin Name: Page for Custom Post Type
Plugin URI: https://github.com/benhuson/page-for-cpt
Description: Specify a page to use for the base URL of a custom post type via your WordPress Reading Settings admin page. May not work with post types that specify their own custom permalink structure.
Author: Ben Huson
Author URI: https://github.com/benhuson/page-for-cpt
Version: 0.6
License: GPLv2
*/

if ( ! defined( 'ABSPATH' ) ) exit;  // Exit if accessed directly.

// Don't load if class already exists
if ( ! class_exists( 'Page_For_CPT' ) ) {

	add_action( 'plugins_loaded', array( 'Page_For_CPT', 'load' ) );

	/**
	 * Page for Custom Post Type Class
	 */
	class Page_For_CPT {

		/**
		 * Plugin File
		 *
		 * @since  0.4
		 *
		 * @var  string
		 */
		private static $file = __FILE__;

		/**
		 * Post Types
		 *
		 * @since  0.1
		 *
		 * @var  array
		 */
		public static $post_types = array();

		/**
		 * Load
		 *
		 * @since  0.4
		 */
		public static function load() {

			add_filter( 'body_class', array( get_class(), 'body_class' ) );
			add_filter( 'get_the_archive_title', array( get_class(), 'get_the_archive_title' ) );
			add_filter( 'get_the_archive_description', array( get_class(), 'get_the_archive_description' ) );
			add_action( 'wp_before_admin_bar_render', array( get_class(), 'toolbar_item' ) );

			if ( self::has_wp_version( '4.4' ) ) {
				add_filter( 'register_post_type_args', array( get_class(), 'register_post_type_args' ), 10, 2 );
			} else {
				add_action( 'registered_post_type', array( get_class(), 'registered_post_type' ), 5, 2 );
			}

			if ( is_admin() ) {
				if ( defined( 'DOING_AJAX' ) && DOING_AJAX ) {
					// Load AJAX functions here if required...
				} else {
					require_once( self::dir() . 'admin/admin.php' );
				}
			}

		}

		/**
		 * The Post
		 *
		 * @since  0.1
		 *
		 * @param   string   $post_type  Post type.
		 * @return  boolean
		 */
		public static function the_post( $post_type ) {

			global $post;

			if ( post_type_exists( $post_type ) ) {

				$page_id = self::get_page_for_post_type( $post_type );

				if ( $page_id > 0 ) {

					if ( 'page' == get_post_type( $page_id ) && 'publish' == get_post_status( $page_id ) ) {
						$post = get_post( $page_id );
						return setup_postdata( $post );
					}

				}

			}

			return false;

		}

		/**
		 * Body Class
		 *
		 * @since  0.1
		 *
		 * @param   array  $classes  Body classes.
		 * @return  array            Body classes.
		 */
		public static function body_class( $classes ) {

			if ( is_post_type_archive() ) {

				foreach ( self::$post_types as $post_type ) {
					if ( is_post_type_archive( $post_type ) ) {

						$page_id = self::get_page_for_post_type( $post_type );

						if ( $page_id > 0 && apply_filters( 'page_for_cpt_use_page_body_classes', false, $post_type ) ) {

							// Add page classes
							$classes[] = 'page';
							$classes[] = 'page-' . $page_id;
							$classes[] = 'page-template-default';

							// Remove general archive classes
							$classes = array_diff( $classes, array( 'archive', 'post-type-archive' ) );

						}

						break;

					}
				}

			}

			return $classes;

		}

		/**
		 * Get The Archive Title
		 *
		 * @since  0.3
		 *
		 * @param   string  $title  Archive title.
		 * @return  string          Filtered archive title.
		 */
		public static function get_the_archive_title( $title ) {

			if ( is_post_type_archive() || is_home() ) {

				if ( is_home() ) {
					$page_id = get_option( 'page_for_posts' );
				} else {
					$page_id = self::get_page_for_post_type( get_queried_object()->name );
				}

				if ( $page_id > 0 ) {
					$title = get_the_title( $page_id );
				}

			}

			return $title;

		}

		/**
		 * Get The Archive Description
		 *
		 * @since  0.3
		 *
		 * @param   string  $description  Archive description.
		 * @return  string                Filtered archive description.
		 */
		public static function get_the_archive_description( $description ) {

			if ( is_post_type_archive() || is_home() ) {

				if ( is_home() ) {
					$page_id = get_option( 'page_for_posts' );
				} else {
					$page_id = self::get_page_for_post_type( get_queried_object()->name );
				}

				if ( $page_id > 0 ) {
					$page = get_post( $page_id );
					$description = apply_filters( 'the_content', $page->post_content );
				}

			}

			return $description;

		}

		/**
		 * Get Post Type for Page
		 *
		 * @since  0.2
		 *
		 * @return  string|false  Post type.
		 */
		public static function get_post_type_for_page() {

			if ( is_page() ) {

				$page_for_cpt = self::get_page_for_cpt_option();
				$post_type = array_search( get_queried_object_id(), $page_for_cpt );

				if ( post_type_exists( $post_type ) ) {
					return $post_type;
				}

			}

			return false;

		}

		/**
		 * Get Page for Post Type
		 *
		 * @since  0.2
		 *
		 * @param   string  $post_type  Post type.
		 * @return  int                 Post ID.
		 */
		public static function get_page_for_post_type( $post_type ) {

			$page_for_cpt = self::get_page_for_cpt_option();

			if ( isset( $page_for_cpt[ $post_type ] ) ) {
				return absint( $page_for_cpt[ $post_type ] );
			}

			return 0;

		}

		/**
		 * Get Page for CPT Option
		 *
		 * @since  0.4
		 *
		 * @return  array  Option data.
		 */
		public static function get_page_for_cpt_option() {

			return (array) get_option( 'page_for_cpt' );

		}

		/**
		 * Get Public Post Types
		 *
		 * @since  0.4
		 *
		 * @return  array  Post types.
		 */
		public static function get_public_post_types() {

			return array_keys( get_post_types( array(
				'public'   => true,
				'_builtin' => false
			) ) );

		}

		/**
		 * Register Post Type Args
		 *
		 * Change the slug for post types.
		 *
		 * @since     0.4
		 * @internal  Called by `register_post_type_args` hook.
		 *
		 * @param   array   $args       Post type args.
		 * @param   string  $post_type  Post type.
		 * @return  array               Post type args.
		 */
		public static function register_post_type_args( $args, $post_type ) {

			// Get all public custom post types.
			$post_types = self::get_public_post_types();

			if ( isset( $args['public'] ) && $args['public'] && ( ! isset( $args['_builtin'] ) || ! $args['_builtin'] ) ) {

				$page_for_cpt = self::get_page_for_post_type( $post_type );

				if ( $page_for_cpt > 0 ) {

					$uri = get_page_uri( $page_for_cpt );

					// If a page is assigned, use that for the rewrite rules.
					if ( ! empty( $uri ) ) {

						$args['has_archive'] = $uri;
						$args['rewrite'] = wp_parse_args( array(
							'slug'       => $uri,
							'with_front' => false
						), (array) $args['rewrite'] );

					}

				}

			}

			return $args;

		}

		/**
		 * Registered Post Type
		 *
		 * Called by the `registered_post_type` hook when a new post type is registered.
		 *
		 * @since  0.3
		 * @internal
		 *
		 * @param  string  $post_type  Post type.
		 * @param  array   $args       Post type arguments.
		 */
		public static function registered_post_type( $post_type, $args ) {

			global $wp_post_types;

			// Get all public custom post types.
			$post_types = self::get_public_post_types();

			if ( in_array( $post_type, $post_types ) ) {

				$page_for_cpt = self::get_page_for_post_type( $post_type );

				if ( $page_for_cpt > 0 ) {

					$uri = get_page_uri( $page_for_cpt );

					// If a page is assigned, use that for the rewrite rules.
					if ( ! empty( $uri ) ) {

						$args->has_archive = $uri;

						$args->rewrite = wp_parse_args( array(
							'slug'       => $uri,
							'with_front' => false
						), (array) $args->rewrite );

					}

					// Rebuild Rewrite Rules
					self::add_post_type_rewrite_rules( $post_type, $args );

					$wp_post_types[ $post_type ] = $args;

				}

			}

		}

		/**
		 * Add Post Type Rewrite Rules
		 *
		 * See the register_post_type() function in `wp-includes/post.php`
		 * https://core.trac.wordpress.org/browser/trunk/src/wp-includes/post.php#L1427
		 *
		 * @since  0.3
		 * @internal
		 *
		 * @param  string  $post_type  Post type.
		 * @param  array   $args       Post type args.
		 */
		public static function add_post_type_rewrite_rules( $post_type, $args ) {

			global $wp_post_types, $wp_rewrite;

			if ( is_admin() || '' != get_option( 'permalink_structure' ) ) {

				if ( $args->has_archive ) {

					$archive_slug = $args->has_archive === true ? $args->rewrite['slug'] : $args->has_archive;

					if ( $args->rewrite['with_front'] ) {
						$archive_slug = substr( $wp_rewrite->front, 1 ) . $archive_slug;
					} else {
						$archive_slug = $wp_rewrite->root . $archive_slug;
					}

					add_rewrite_rule( "{$archive_slug}/?$", "index.php?post_type=$post_type", 'top' );

					if ( $args->rewrite['feeds'] && $wp_rewrite->feeds ) {
						$feeds = '(' . trim( implode( '|', $wp_rewrite->feeds ) ) . ')';
						add_rewrite_rule( "{$archive_slug}/feed/$feeds/?$", "index.php?post_type=$post_type" . '&feed=$matches[1]', 'top' );
						add_rewrite_rule( "{$archive_slug}/$feeds/?$", "index.php?post_type=$post_type" . '&feed=$matches[1]', 'top' );
					}

					if ( $args->rewrite['pages'] ) {
						add_rewrite_rule( "{$archive_slug}/{$wp_rewrite->pagination_base}/([0-9]{1,})/?$", "index.php?post_type=$post_type" . '&paged=$matches[1]', 'top' );
					}

				}

				$permastruct_args = $args->rewrite;
				$permastruct_args['feed'] = $permastruct_args['feeds'];
				add_permastruct( $post_type, "{$args->rewrite['slug']}/%$post_type%", $permastruct_args );

			}

		}

		/**
		 * Toolbar Item
		 *
		 * Add back "Edit Page" tolbar item when viewing a post type archive
		 * which has an associated page.
		 *
		 * @since     0.6
		 * @internal  Private. Called via `wp_before_admin_bar_render` actions.
		 */
		public static function toolbar_item() {

			global $wp_admin_bar;

			$current_object = get_queried_object();

			if ( ! is_admin() && is_post_type_archive() && ! empty( $current_object ) ) {

				$page_id = self::get_page_for_post_type( $current_object->name );
				$post_type_object = get_post_type_object( 'page' );

				if ( ! empty( $page_id )
					&& $post_type_object
					&& current_user_can( 'edit_post', $page_id )
					&& $post_type_object->show_in_admin_bar
					&& $edit_post_link = get_edit_post_link( $page_id )
					) {

					$wp_admin_bar->add_menu( array(
						'id'    => 'edit',
						'title' => $post_type_object->labels->edit_item,
						'href'  => $edit_post_link
					) );

				}

			}

		}

		/**
		 * Has WordPress Version
		 *
		 * @since  0.4
		 *
		 * @param   string  $version  Version in format `0.0.0`
		 * @return  boolean           True if WordPress version is same or higher.
		 */
		public static function has_wp_version( $version ) {

			global $wp_version;

			return version_compare( $wp_version, $version ) >= 0;

		}

		/**
		 * Plugin Basename
		 *
		 * @since  0.4
		 *
		 * @return  string  Plugin basename.
		 */
		public static function basename() {

			return plugin_basename( self::$file );

		}

		/**
		 * Plugin Sub Directory
		 *
		 * @since  0.4
		 *
		 * @return  string  Plugin folder name.
		 */
		public static function sub_dir() {

			return '/' . str_replace( basename( self::$file ), '', self::basename() );

		}

		/**
		 * Plugin URL
		 *
		 * @since  0.4
		 *
		 * @return  string  Plugin directory URL.
		 */
		public static function url() {

			return plugins_url( self::sub_dir() );

		}

		/**
		 * Plugin Directory
		 *
		 * @since  0.4
		 * 
		 * @return  string  Plugin directory path.
		 */
		public static function dir() {

			return plugin_dir_path( self::$file );

		}

	}

}
